src/System.Security.Claims/src/System/Security/Claims/ClaimsIdentity.cs

```diff
@@ -0,0 +1,970 @@
+        /// <param name="authenticationType">The type of authentication used.</param>
+        /// <param name="nameType">The <see cref="Claim.Type"/> used when obtaining the value of <see cref="ClaimsIdentity.Name"/>.</param>
+        /// <param name="roleType">The <see cref="Claim.Type"/> used when performing logic for <see cref="ClaimsPrincipal.IsInRole"/>.</param>
+        /// <remarks>If 'identity' is a <see cref="ClaimsIdentity"/>, then there are potentially multiple sources for AuthenticationType, NameClaimType, RoleClaimType.
+        /// <para>Priority is given to the parameters: authenticationType, nameClaimType, roleClaimType.</para>
+        /// <para>All <see cref="Claim"/>s are copied into this instance in a <see cref="List{Claim}"/>. Each Claim is examined and if Claim.Subject != this, then Claim.Clone(this) is called before the claim is added.</para>
+        /// <para>Any 'External' claims are ignored.</para>
+        /// </remarks>
+        /// <exception cref="InvalidOperationException">if 'identity' is a <see cref="ClaimsIdentity"/> and <see cref="ClaimsIdentity.Actor"/> results in a circular refrence back to 'this'.</exception>
+        public ClaimsIdentity(IIdentity identity, IEnumerable<Claim> claims, string authenticationType, string nameType, string roleType)
+        {
+            // TODO - brentsch, should we have another constructor that takes ClaimsIdentity thereby bypassing the cast below?
+            ClaimsIdentity claimsIdentity = identity as ClaimsIdentity;
+
+            _authenticationType = !string.IsNullOrWhiteSpace(authenticationType) ? authenticationType : (identity != null ? identity.AuthenticationType : null);
+            _nameClaimType = !string.IsNullOrEmpty(nameType) ? nameType : (claimsIdentity != null ? claimsIdentity._nameClaimType : DefaultNameClaimType);
+            _roleClaimType = !string.IsNullOrEmpty(roleType) ? roleType : (claimsIdentity != null ? claimsIdentity._roleClaimType : DefaultRoleClaimType);
+
+            if (claimsIdentity != null)
+            {
+                _label = claimsIdentity._label;
+                _bootstrapContext = claimsIdentity._bootstrapContext;
+
+                if (claimsIdentity.Actor != null)
+                {
+                    //
+                    // Check if the Actor is circular before copying. That check is done while setting
+                    // the Actor property and so not really needed here. But checking just for sanity sake
+                    //
+                    if (!IsCircular(claimsIdentity.Actor))
+                    {
+                        _actor = claimsIdentity.Actor;
+                    }
+                    else
+                    {
+                        throw new InvalidOperationException(SR.GetResourceString("InvalidOperationException_ActorGraphCircular", "Actor cannot be set so that circular directed graph will exist chaining the subjects together."));
```

Comment:
We should use the same mechanism we use elsewhere in corefx for resources, e.g. ```SR.InvalidOperationException_ActorGraphCircular```.